wagtail_newsletter/campaign_backends/mailchimp.py

```diff
@@ -1,32 +1,80 @@
-from typing import Any
+import logging
+
+from copy import copy
+from dataclasses import dataclass
+from typing import Any, NoReturn, Optional, cast
 
 from django.conf import settings
 from django.core.exceptions import ImproperlyConfigured
+from django.utils.functional import cached_property
 from mailchimp_marketing import Client
 from mailchimp_marketing.api_client import ApiClientError
 
 from ..audiences import (
     Audience,
     AudienceSegment,
 )
-from . import CampaignBackend
+from ..models import NewsletterRecipientsBase
+from . import Campaign, CampaignBackend, CampaignBackendError
+
+
+CAMPAIGN_STATUS_DRAFT = "save"
+
+logger = logging.getLogger(__name__)
+
+
+def log_and_raise(error: ApiClientError, message: str, **kwargs) -> NoReturn:
+    kwargs["status_code"] = error.status_code
+    kwargs["text"] = error.text
+    logger.exception(
+        f"{message}: {', '.join(f'{key}=%r' for key in kwargs.keys())}",
+        *kwargs.values(),
+    )
+    raise CampaignBackendError(message) from error
+
+
+def require_setting(name):
+    value = getattr(settings, name, None)
+    if value is None:
+        raise ImproperlyConfigured(f"{name} is not set")
+    return value
+
+
+@dataclass
+class MailchimpCampaign(Campaign):
+    backend: "MailchimpCampaignBackend"
+    id: str
+    web_id: str
+    status: str
+
+    @property
+    def sent(self):
+        return self.status != CAMPAIGN_STATUS_DRAFT
+
+    @property
+    def url(self):
+        server = self.backend.client.api_client.server
+        base_url = f"https://{server}.admin.mailchimp.com"
+
+        if self.sent:
+            return f"{base_url}/reports/summary?id={self.web_id}"
+
+        else:
+            return f"{base_url}/campaigns/edit?id={self.web_id}"
 
 
 class MailchimpCampaignBackend(CampaignBackend):
     name = "Mailchimp"
 
-    def __init__(self):
-        self.client = Client()
-        self.client.set_config(self.get_client_config())
+    @cached_property
+    def client(self):
+        client = Client()
+        client.set_config(self.get_client_config())
+        return client
```

Comment:
I've changed `client` to a `@cached_property` so that any `ImproperlyConfigured` errors from `get_client_config` are thrown later, when API calls are actually made. This means we can always instantiate the backend to get its `name` attribute, which is needed to render the editor panel.

Reply:
That seems to be a sensible use of a lazy-instanciation indeed! :+1: :slightly_smiling_face: